tests/phpunit/CRM/Core/BAO/SettingTest.php

```diff
@@ -180,4 +180,18 @@ public static function _testOnChange_onChangeExample($oldValue, $newValue, $meta
     $_testOnChange_hookCalls['metadata'] = $metadata;
   }
 
+  /**
+   * Test to set isProductionEnvironment
+   *
+   */
+  public function testSetCivicrmEnvironment() {
+    CRM_Core_BAO_Setting::setItem('Staging', CRM_Core_BAO_Setting::DEVELOPER_PREFERENCES_NAME, 'environment');
+    $values = CRM_Core_BAO_Setting::getItem(CRM_Core_BAO_Setting::DEVELOPER_PREFERENCES_NAME, 'environment');
+    $this->assertEquals('Staging', $values);
+
+    define('CIVICRM_ENVIRONMENT', 'Development');
```

Comment:
(a) It's not safe for tests to call `define()` (at least, given the standard headless test environment). Trivially, suppose this test runs twice in a row. The second time, it will fail. Less trivially, it makes the test environment harder to predict. Compare:

 * When you run `CRM_AllTests`, any earlier tests (like `CRM/Activity/**Test.php`) run in an open-ended environment. (The default appears to be `Production`.)
 * When you run `CRM_AllTests`, any later tests (like `CRM/Utils/**Test.php`) are unalterably pegged to the `Development` environment.
 * When you run any test by itself (like `CRM/Utils/**Test.php`), you again have an open-ended environment (where the default is `Production`).
 * Consequently, tests will *some of the time* run in `Production` and *some of the time* in `Development`.

(b) Does it actually help to provide special handling for this setting (`define()`, `CRM_Core_Config::environment()`) rather than vanilla settings (`$civicrm_setitings`, `Civi::settings()->get('environment')`)?

 * For example, if the reason for the `define()` is that you want to be able to hard-code a value in `civicrm.settings.php`... then that's already handled by setting `$civicrm_settings['domain']['environment']`. (Note: Recall that in 4.7 you don't need to index by the `DEVELOPER_PREFERENCES_NAME` or whatever -- they're all just aliases for `domain`.) 
 * On the other hand, if the setting is *boot critical* (meaning that it must be available as part of the early initialization of the settings/extensions/container system), then maybe it is needed. (But I'd really curious to know why it's boot-critical -- nothing pops while reading.)

Reply:
My immediate thinking here is that what the test here is trying to prove is that if there is a value set in civicrm.settings.php that it overrides everything else. so you have a method set using essentially `Civi::settings()->set` to set the value to staging but then someone is then overriding that by setting Development in `civicrm.settings.php` However i'm thinking that the style at https://github.com/civicrm/civicrm-core/pull/8724/files#diff-d392832fe610e31e06039270b3a100e6R546 seems to be more robust way of testing `civicrm.settings.php`

Reply:
@seamuslee001 check Tim's comment about getMandatory - it makes me happy to see that